iothub/service/src/ConnectionLeaseTimeoutHandler.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices
+{
+    /// <summary>
+    /// HttpClient delegating handler for periodically revoking the connection lease after a specified amount of time. This allows cached HttpClients to periodically do DNS lookups
+    /// which keeps the client communicating with the correct service endpoint in failover scenarios.
+    /// </summary>
+    internal class ConnectionLeaseTimeoutHandler : DelegatingHandler
+    {
+        private readonly Stopwatch _stopwatch;
+        private readonly int _leaseTimeoutMilliseconds;
+        private SemaphoreSlim _timeoutCheckSemaphore;
+
+        internal ConnectionLeaseTimeoutHandler(int leaseTimeoutMilliseconds, HttpClientHandler innerHandler) : base()
+        {
+            _stopwatch = Stopwatch.StartNew();
+            _timeoutCheckSemaphore = new SemaphoreSlim(1, 1);
+            _leaseTimeoutMilliseconds = leaseTimeoutMilliseconds;
+            InnerHandler = innerHandler;
+        }
+
+        protected async override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            // By default, singleton/static HttpClient instances will never do a DNS lookup after the first request, which leads to some
+            // bugs such as https://github.com/Azure/azure-iot-sdk-csharp/issues/1865
+
+            // Special case for lease timeout is when the value is set to a negative value. In those instances, the lease timeout is infinite,
+            // so there is no reason to ever set the ConnectionClose header. As such, there is no need to wait for the semaphore to be available.
+            if (_leaseTimeoutMilliseconds >= 0)
```

Comment:
Shouldn't you check to see if the timer elapsed exceeds the threshold before getting the semaphore too? That way you don't get the semaphore on each request.

You'll need to check the value again after the lock, of course, to handle concurrent requests that hit the initial condition and try to get the semaphore.